src/commands/synthetics/utils.ts

```diff
@@ -292,148 +281,85 @@ export const waitForResults = async (
-      }
-    }
-
-    if (pollingDuration >= maxPollingTimeout) {
-      break
-    }
-
-    let polledResults: PollResult[]
+  const maxPollingDate = Date.now() + options.maxPollingTimeout
+  const emittedResultIndexes = new Set<number>()
+  const processBatch = async () => {
     try {
-      polledResults = (await api.pollResults([...inProgressTriggers.values()].map((tr) => tr.result_id))).results
-    } catch (error) {
-      if (is5xxError(error) && !options.failOnCriticalErrors) {
-        polledResults = []
-        for (const triggerResult of inProgressTriggers) {
-          inProgressTriggers.delete(triggerResult)
-          const test = getTest(triggerResult.public_id)
-          const result = createFailingResult(
-            ERRORS.ENDPOINT,
-            triggerResult,
-            getLocation(triggerResult.location, test),
-            test
-          )
-          result.passed = hasResultPassed(result.result, options.failOnCriticalErrors!!, options.failOnTimeout!!)
-          results.push(result)
-        }
-      } else {
-        throw error
-      }
-    }
-
-    for (const polledResult of polledResults) {
-      if (polledResult.result.eventType === 'finished') {
-        const triggeredResult = [...inProgressTriggers.values()].find((r) => r.result_id === polledResult.resultID)
-        if (triggeredResult) {
-          inProgressTriggers.delete(triggeredResult)
-          const test = getTest(triggeredResult.public_id)
-          const result: Result = {
-            enrichment: polledResult.enrichment,
-            location: getLocation(triggeredResult.location, test),
-            passed: hasResultPassed(polledResult.result, options.failOnCriticalErrors!!, options.failOnTimeout!!),
-            result: polledResult.result,
-            resultId: polledResult.resultID,
-            test: deepExtend(test, polledResult.check),
-            timestamp: polledResult.timestamp,
-          }
-          results.push(result)
+      const currentBatch = await api.getBatch(trigger.batch_id)
+      for (const [index, result] of currentBatch.results.entries()) {
+        if (result.status !== 'in_progress' && !emittedResultIndexes.has(index)) {
+          emittedResultIndexes.add(index)
           reporter.resultReceived(result)
         }
       }
-    }
 
-    if (!inProgressTriggers.size) {
-      break
+      return currentBatch
+    } catch (e) {
+      throw new EndpointError(`Failed to get batch: ${formatBackendErrors(e)}\n`, e.response?.status)
     }
+  }
 
+  let batch = await processBatch()
+  // In theory polling the batch is enough, but in case something goes wrong backend-side
+  // let's add a check to ensure it eventually times out.
+  let hasExceededMaxPollingDate = Date.now() >= maxPollingDate
+  while (batch.status === 'in_progress' && !hasExceededMaxPollingDate) {
+    batch = await processBatch()
     await wait(POLLING_INTERVAL)
+    hasExceededMaxPollingDate = Date.now() >= maxPollingDate
   }
```

Comment:
I wish I could to avoid the duplication, but I need the latest value here: https://github.com/DataDog/datadog-ci/pull/551/files/19f5c296c64f60e2b0cc1cadcf22ef9c891b2406#diff-741a985551714cec0c6f7a6423d704525619e8f57db6c4f7c685d95d88a8d0cfR341